File: node/fe/ui/pages/trading.js
'use strict';

/**
 * The live DEX trading screen, read by LABEL.
 *
 * The page is a React app with hashed class names and no test IDs, so the
 * reliable anchor is the text a trader reads: the header runs
 *
 *   <PAIR>/USD | [POOL] | $ price | change% | 24H VOLUME | $ vol |
 *   OPEN INTEREST (L%/S%) | $ long | / | $ short | NET RATE / 1H | rate | / | rate
 *
 * Everything below extracts from that sequence of lines, so a redesign that
 * keeps the labels keeps the tests, and a redesign that drops a label fails as
 * ScreenNotReady rather than reading a neighbouring number as the wrong thing.
 *
 * Readiness is by LABEL WITH A NUMBER, never by time: the app takes 10-25 s to
 * fill in, and a fixed sleep either wastes it or reads a blank. If the labelled
 * number never arrives the page raises ScreenNotReady, which the step layer
 * turns into Blocked -- a screen that did not load has not shown a wrong price.
 *
 * Market selection goes through the pair dropdown; the URL alone does not
 * switch markets (measured).
 */

const URL = 'https://app.gmx.io/#/trade';

class ScreenNotReady extends Error {
  constructor(message) {
    super(message);
    this.name = 'ScreenNotReady';
  }
}

/** "$ 6.0m" -> { value: 6000000, unit: 100000 }; "$ 170.4k" -> { 170400, 100 }; "$ 2,522.99" -> { 2522.99, 0.01 } */
function parseMoney(text) {
  const m = String(text).replace(/[$,\s]/g, '').match(/^([-+]?\d+(?:\.\d+)?)([kmb])?$/i);
  if (!m) return null;
  const n = Number(m[1]);
  const mult = { k: 1e3, m: 1e6, b: 1e9 }[(m[2] || '').toLowerCase()] || 1;
  const decimals = (m[1].split('.')[1] || '').length;
  // the resolution the screen rounded to: one unit in the last shown digit
  const unit = mult / 10 ** decimals;
  return { value: n * mult, unit };
}

/** "- 0.0010%" -> -0.000010 (a fraction); "+ 0.0005%" -> 0.000005 */
function parsePercent(text) {
  const m = String(text).replace(/\s/g, '').match(/^([-+])?(\d+(?:\.\d+)?)%$/);
  if (!m) return null;
  return (m[1] === '-' ? -1 : 1) * Number(m[2]) / 100;
}

/** "(53%/47%)" -> { long: 53, short: 47 } */
function parseSplit(text) {
  const m = String(text).match(/\((\d+)%\/(\d+)%\)/);
  return m ? { long: Number(m[1]), short: Number(m[2]) } : null;
}

async function bodyLines(page) {
  return page.evaluate(() =>
    document.body.innerText.split('\n').map((s) => s.trim()).filter(Boolean));
}

class TradingPage {
  constructor(page) {
    this.page = page;
  }

  async open() {
    try {
      await this.page.goto(URL, { waitUntil: 'domcontentloaded', timeout: 60_000 });
    } catch (err) {
      throw new ScreenNotReady('could not open ' + URL + ' -- ' + err.message.split('\n')[0]);
    }
  }

  /**
   * Choose a market through the pair dropdown, then wait for its header.
   *
   * The app re-renders the header while prices stream in, so a click can land
   * on a node that was just replaced. Three attempts; any interaction failure
   * is ScreenNotReady, because a dropdown that would not open has not shown a
   * wrong number.
   */
  async selectMarket(pair, attempts = 3) {
    let lastErr = null;
    for (let i = 0; i < attempts; i++) {
      try {
        await this.waitForHeader();
        const current = await this.readHeader();
        if (current.pair === pair) return;

        const opener = this.page.getByText(new RegExp('^' + current.pair.replace('/', '\\/') + '$')).first();
        await opener.click({ timeout: 15_000 });
        const option = this.page.getByText(new RegExp('^' + pair.replace('/', '\\/') + '$')).first();
        await option.click({ timeout: 15_000 });
        await this.waitForHeader(pair);
        return;
      } catch (err) {
        if (err instanceof ScreenNotReady) throw err;
        lastErr = err;
        await this.page.keyboard.press('Escape').catch(() => {});
        await this.page.waitForTimeout(1_500);
      }
    }
    throw new ScreenNotReady('could not select market ' + pair + ' after ' + attempts +
      ' attempts -- ' + String(lastErr && lastErr.message).split('\n')[0]);
  }

  /** Wait until the header shows a number after 24H VOLUME (and the wanted pair, if given). */
  async waitForHeader(pair = null, timeoutMs = 60_000) {
    const started = Date.now();
    while (Date.now() - started < timeoutMs) {
      const lines = await bodyLines(this.page);
      const i = lines.findIndex((l) => /^24H VOLUME$/i.test(l));
      const pairLine = lines.slice(0, i > 0 ? i : 0).reverse().find((l) => /\/USD$/.test(l));
      if (i > 0 && /\d/.test(lines[i + 1] || '') && (!pair || pairLine === pair)) return;
      await this.page.waitForTimeout(1_000);
    }
    throw new ScreenNotReady('the trading header never showed a 24H VOLUME figure' +
      (pair ? ' for ' + pair : '') + ' within ' + timeoutMs / 1000 + 's');
  }

  /**
   * Everything in the header, parsed. Numbers keep the resolution the screen
   * rounded to (`unit`), so a comparison can say "within what was displayed".
   */
  async readHeader() {
    const lines = await bodyLines(this.page);
    const iVol = lines.findIndex((l) => /^24H VOLUME$/i.test(l));
    if (iVol < 0) throw new ScreenNotReady('no 24H VOLUME label on the page');

    const before = lines.slice(Math.max(0, iVol - 6), iVol);
    const pair = [...before].reverse().find((l) => /\/USD$/.test(l)) || null;
    const pool = before.find((l) => /^\[.+\]$/.test(l)) || null;
    const priceText = before.find((l) => /^\$\s?[\d,]+(\.\d+)?$/.test(l)) || null;
    const changeText = before.find((l) => /^[-+]?\d+(\.\d+)?%$/.test(l)) || null;

    const iOi = lines.findIndex((l, k) => k > iVol && /^OPEN INTEREST/i.test(l));
    const iRate = lines.findIndex((l, k) => k > iVol && /^NET RATE/i.test(l));
    if (iOi < 0 || iRate < 0) throw new ScreenNotReady('header is missing OPEN INTEREST or NET RATE');

    // "OPEN INTEREST (50%/50%)", "$ 6.0m", "/", "$ 6.0m"
    const oiLong = parseMoney(lines[iOi + 1]);
    const oiShort = parseMoney(lines[iOi + 3]);
    // "NET RATE / 1H", "- 0.0010%", "/", "+ 0.0005%"
    const rateLong = parsePercent(lines[iRate + 1]);
    const rateShort = parsePercent(lines[iRate + 3]);

    return {
      pair,
      pool,
      price: priceText ? parseMoney(priceText) : null,
      change: changeText ? parsePercent(changeText) : null,
      volume24h: parseMoney(lines[iVol + 1]),
      oiSplit: parseSplit(lines[iOi]),
      oiLong,
      oiShort,
      netRateLong: rateLong,
      netRateShort: rateShort,
      raw: lines.slice(iVol - 4, iRate + 4),
    };
  }
}

module.exports = { TradingPage, ScreenNotReady, parseMoney, parsePercent, parseSplit, URL };

File: python/fe/ui/pages/trading.py
"""The live DEX trading screen, read by LABEL.

Mirror of ``node/fe/ui/pages/trading.js`` -- see there for why labels, why
readiness is "a label with a number", and why the market is chosen through
the dropdown rather than the URL.
"""

from __future__ import annotations

import re
import time

URL = "https://app.gmx.io/#/trade"


class ScreenNotReady(Exception):
    """The screen did not load, or never filled in the labelled figure."""


def parse_money(text: str) -> dict | None:
    """'$ 6.0m' -> {value: 6000000, unit: 100000}; '$ 2,522.99' -> {2522.99, 0.01}."""
    m = re.match(r"^([-+]?\d+(?:\.\d+)?)([kmb])?$", re.sub(r"[$,\s]", "", str(text)), re.I)
    if not m:
        return None
    n = float(m.group(1))
    mult = {"k": 1e3, "m": 1e6, "b": 1e9}.get((m.group(2) or "").lower(), 1)
    decimals = len(m.group(1).split(".")[1]) if "." in m.group(1) else 0
    return {"value": n * mult, "unit": mult / 10 ** decimals}


def parse_percent(text: str) -> float | None:
    """'- 0.0010%' -> -0.000010 (a fraction)."""
    m = re.match(r"^([-+])?(\d+(?:\.\d+)?)%$", re.sub(r"\s", "", str(text)))
    if not m:
        return None
    return (-1 if m.group(1) == "-" else 1) * float(m.group(2)) / 100


def parse_split(text: str) -> dict | None:
    m = re.search(r"\((\d+)%/(\d+)%\)", str(text))
    return {"long": int(m.group(1)), "short": int(m.group(2))} if m else None


def _body_lines(page) -> list[str]:
    return page.evaluate("() => document.body.innerText.split('\\n').map(s => s.trim()).filter(Boolean)")


class TradingPage:
    def __init__(self, page) -> None:
        self.page = page

    def open(self) -> None:
        try:
            self.page.goto(URL, wait_until="domcontentloaded", timeout=60_000)
        except Exception as err:  # playwright raises its own Error subclasses
            raise ScreenNotReady(f"could not open {URL} -- {str(err).splitlines()[0]}") from err

    def select_market(self, pair: str, attempts: int = 3) -> None:
        """Choose a market through the pair dropdown, then wait for its header.

        The app re-renders the header while prices stream in, so a click can
        land on a node that was just replaced. Three attempts; any interaction
        failure is ScreenNotReady -- a dropdown that would not open has not
        shown a wrong number.
        """
        last_err: Exception | None = None
        for _ in range(attempts):
            try:
                self.wait_for_header()
                current = self.read_header()
                if current["pair"] == pair:
                    return
                self.page.get_by_text(re.compile("^" + re.escape(current["pair"]) + "$")).first.click(timeout=15_000)
                self.page.get_by_text(re.compile("^" + re.escape(pair) + "$")).first.click(timeout=15_000)
                self.wait_for_header(pair)
                return
            except ScreenNotReady:
                raise
            except Exception as err:  # playwright's own error classes
                last_err = err
                try:
                    self.page.keyboard.press("Escape")
                except Exception:
                    pass
                self.page.wait_for_timeout(1_500)
        raise ScreenNotReady(
            f"could not select market {pair} after {attempts} attempts -- {str(last_err).splitlines()[0]}"
        )

    def wait_for_header(self, pair: str | None = None, timeout_s: float = 60) -> None:
        started = time.monotonic()
        while time.monotonic() - started < timeout_s:
            lines = _body_lines(self.page)
            i = next((k for k, l in enumerate(lines) if re.match(r"^24H VOLUME$", l, re.I)), -1)
            pair_line = next((l for l in reversed(lines[:i]) if l.endswith("/USD")), None) if i > 0 else None
            if i > 0 and re.search(r"\d", lines[i + 1] if i + 1 < len(lines) else "") and (not pair or pair_line == pair):
                return
            self.page.wait_for_timeout(1_000)
        raise ScreenNotReady(
            "the trading header never showed a 24H VOLUME figure"
            + (f" for {pair}" if pair else "") + f" within {timeout_s:.0f}s"
        )

    def read_header(self) -> dict:
        lines = _body_lines(self.page)
        i_vol = next((k for k, l in enumerate(lines) if re.match(r"^24H VOLUME$", l, re.I)), -1)
        if i_vol < 0:
            raise ScreenNotReady("no 24H VOLUME label on the page")

        before = lines[max(0, i_vol - 6):i_vol]
        pair = next((l for l in reversed(before) if l.endswith("/USD")), None)
        pool = next((l for l in before if re.match(r"^\[.+\]$", l)), None)
        price_text = next((l for l in before if re.match(r"^\$\s?[\d,]+(\.\d+)?$", l)), None)
        change_text = next((l for l in before if re.match(r"^[-+]?\d+(\.\d+)?%$", l)), None)

        i_oi = next((k for k, l in enumerate(lines) if k > i_vol and re.match(r"^OPEN INTEREST", l, re.I)), -1)
        i_rate = next((k for k, l in enumerate(lines) if k > i_vol and re.match(r"^NET RATE", l, re.I)), -1)
        if i_oi < 0 or i_rate < 0:
            raise ScreenNotReady("header is missing OPEN INTEREST or NET RATE")

        return {
            "pair": pair,
            "pool": pool,
            "price": parse_money(price_text) if price_text else None,
            "change": parse_percent(change_text) if change_text else None,
            "volume24h": parse_money(lines[i_vol + 1]),
            "oiSplit": parse_split(lines[i_oi]),
            "oiLong": parse_money(lines[i_oi + 1]),
            "oiShort": parse_money(lines[i_oi + 3]),
            "netRateLong": parse_percent(lines[i_rate + 1]),
            "netRateShort": parse_percent(lines[i_rate + 3]),
            "raw": lines[i_vol - 4:i_rate + 4],
        }
